### ISSUE TITLE
Correctives?

### ISSUE THREAD

**Issue by illmillrig** [blurstudio/Simplex#65]
a quick grep of the word 'Corrective' and the resulting methods leads me to believe that Simplex no longer handles corrective shapes?

**Comment by tbttfox** [comment]
Correctives have (so far) been handled by scripts not associated with simplex. To the solver, a corrective is just another shape. It's just a shape that happens to interact with other deformations.
It is, however, something I've meant to add to the extra tools, but I've never gotten around to it. Mostly because we have scripts to do it, the scripts we have are XSI specific, and I personally haven't put much work into consolidating the toolset for XSI.

**Comment by tbttfox** [comment]
That said, there's a really neat way of building correctives over top of a smooth skin (so long as it's only smooth skin). 
First, pose the character, and store the points.
Then you add a deformation of 1 unit in X to each point **before** the skin, store the result, and undo.
Then add a deformation of 1 unit in Y, store that, undo, Z, store that, undo.
You end up with Pose, 1X*Pose, 1Y*Pose, 1Z*Pose.  Subtract the Pose from the 3 with unit deforms and you get 3 sets of vectors. Along with the original pose, you can use them to directly construct a matrix per point. Then just invert those matrices out from under each of your shape's points to give you the corrective deltas.

**Comment by illmillrig** [comment]
lol, I had to read it a few times, but I get it. That's really, really cool.  Cheers!

**Comment by illmillrig** [comment]
okay, quick question about this.  the vectors that are being created from subtracting the pose from the 3 vectors; 1X*Pose, 1Y*Pose, 1Z*Pose, are those being normalized? is that what you mean by "with unit deforms"? or does the resulting matrix at each point need unnormalized vectors for the first 3 rows?

**Comment by tbttfox** [comment]
Oh, I only said "unit vectors" because the xform added before the skin was 1 unit.
The vectors you get at the end need to stay unnormalized for everything to work properly.

**Comment by illmillrig** [comment]
ah gotcha.  okay one more thing :)

the shape points that you're then multiplying by those inverted matrices. are those points in worldspace? or are they vectors created from subtracting the skinned points from the shape points?

**Comment by tbttfox** [comment]
The idea is you come up with a matrix per deformed point, and then get your shape's point in the space of that matrix. So you're going from the shape's worldspace to the point's local space.  Then, you could *technically* multiply those local space vectors by any per-point matrix to transfer that shape to other shapes. But we're going for that shape in rest-space, and those matrices are just the position offset of each vertex, so you can simplify things and just add the vector to the undeformed point position.

tl;dr You input worldspace points, and get back offset vectors.

**Comment by illmillrig** [comment]
ahhhh, okay.  that makes sense now.  Many Thanks!

**Comment by illmillrig** [comment]
rough first draft, but seems to work well enough
https://gist.github.com/illmillrig/e104bb7a48d7c583ed0be1efa30af995
